### ISSUE TITLE
Lecture eleven-Decision Tree, Random Forests

### ISSUE THREAD

**Issue by jimjinyan** [jimjinyan/Go-Just-Go#25]
### Decision Tree
A decision tree is a flowchart-like structure in which each internal node represents a "test" on an attribute (e.g. whether a coin flip comes up heads or tails), each branch represents the outcome of the test, and each leaf node represents a class label (decision taken after computing all attributes). The paths from root to leaf represent classification rules.

- Fast training
- Fast prediction
- Easy to understand
- Easy to interpret

Decision Tree - idea
- found the threshold in feature space (for example in 2 featuers space)
    - Root node, Internal nodes and Leaf or terminal
- query only one feature at one time (for example x_1), which this is a boundary for all the followint queries
    - is a straight line by query one feature
        - more deep tree for complex lines(like diagonal line)
        - Veary fast(high perform); intuitive; actual category, no need for scale
- reach to certian critiria to stop and then assign class to the leaf, whcih is equal to assign a class to a cell

Decision Tree - prediction
- run all the questions by features
    - easy for multiple calsses

Decision Tree - train
- which feasure whould query
- The threshold to choose --Leaf Node purity
- Take a vote to decide category
- Hunt's Algorithm
    - step 1: if all the records in $D_t$ belong to the same class $y_t$, then $t$ is a leaf node labeled as $y_t$
    - step 2: if $D_t$ contains records that belong to more than one class, an attribute test condition is selected to partition the records into smaller subsets. 
The algorithm is recursively applied to each child node. 
    - Hunt's algorithm works if eery combination of attribute calues is present in the training data and each combination has a unique class label.

![split](https://user-images.githubusercontent.com/39241392/54474398-239a1f80-481f-11e9-853b-c24b51954389.png)

How to measure:
Measure the information gain(IG) at each split. The objective is to maximize the IG$$IG(D_p,f)=I(D_p)-\sum_{j=1}^{m}\frac{N_j}{N}I(D_j)$$
Let$p(i|t)$ denote the fraction of records belonging to class i at a given node t.
- The measures developed for selecting the best split are often based on the degree of impurity of the cild nodes (Impurity measures). The smaller the better.
    - Classification Error
        * $I_E=\ 1-max({p(i|t)})$ 
    - Gini Impurity
        * $I_G(t)=1-\sum_{i=1}^{c}p(i|t)^2$
    - Entropy
        * $I_H(t) = -\sum_{i=1}^{c}p(i|t)log_2p(i|t)$
- To determine how well a test condition performs, we need to compare the degree of impurity of the parent node with the degree of impurity of the cild nodes. The large the different (information gain) , the better the test condition. $$\Delta=I(parent)-\sum_{j=1}^{k}\frac{N(v_j)}{N}I(v_j),$$

- What does optimal tree mean?--it means the depth of tree

How about overfit:
1. restrict the test conditons to binary splits only
2. Gain ratio$$Gain\ ratio = \frac{\Delta_info}{Split\ Info},\ \ \ Split\ Info=-\sum_{i=1}^{k}P(v_i)log_2P(v_i)$$
3. Ensemble Methods
    - learn muliple trees by bootstrap
        - Take N data points
        - Draw N times with replacment
        - Can not use bootstrap to evaluate the test performance
        - Bagging--Bootstrap aggregating
            - Consensus --majority vote or average

[Good refference](https://www-users.cs.umn.edu/~kumar001/dmbook/ch4.pdf)

### Random Forest
builds upon the bagging **and ** node splits calculated from **random feature** subest
